orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -0,0 +1,255 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/core/graph/pipeline_transformer.h"
+
+using namespace onnxruntime::common;
+
+namespace onnxruntime {
+namespace training {
+
+void GetPipelineSendOutput(const Graph& graph, std::string& loss_name) {
+  for (auto& node : graph.Nodes()) {
+    if (!node.OpType().compare("Send")) {
+      // send op should always have an output, which is the OutputSignal.
+      loss_name = node.OutputDefs()[0]->Name();
+      return;
+    }
+  }
+}
+
+bool IsBackward(Node& node) {
+  return (node.Description() == "Backward pass");
+}
+
+void AddInputEvent(Graph& graph, const std::string& op_name,
+                   bool is_forward,
+                   std::vector<NodeArg*>& input_args,
+                   std::vector<std::string>& new_input_names) {
+  ONNX_NAMESPACE::TypeProto event_type_proto;
+  event_type_proto.mutable_tensor_type()->set_elem_type(ONNX_NAMESPACE::TensorProto_DataType_INT64);
+
+  auto event_id_name = graph.GenerateNodeArgName(op_name + (is_forward ? "_fw" : "_bw") + "_event_id");
+  auto& event_id = graph.GetOrCreateNodeArg(event_id_name, &event_type_proto);
+  new_input_names.push_back(event_id_name);
+
+  input_args.push_back(&event_id);
+}
+
+void AddRecordBackward(Graph& graph, Node* send_bw, std::vector<std::string>& new_input_names) {
+  // gradient graph can contain some dangling leaf nodes. Add them all to WaitEvent
+  // backward node's input.
+  auto add_leaf_node = [&](std::vector<NodeArg*>& input_args) {
+    for (auto& node : graph.Nodes()) {
+      if (!IsBackward(node)) {
+        // only check backward node
+        continue;
+      }
+      bool find_consumer_nodes = false;
+      std::vector<NodeArg*>& outputs = node.MutableOutputDefs();
+      for (auto& output : outputs) {
+        std::vector<const Node*> consumer_nodes = graph.GetConsumerNodes(output->Name());
+        if (consumer_nodes.size() > 0) {
+          find_consumer_nodes = true;
+          break;
+        }
+      }
+      if (!find_consumer_nodes && outputs.size() > 0) {
+        input_args.push_back(outputs[0]);
+      }
+    }
+  };
+
+  std::vector<NodeArg*> input_args;
+  AddInputEvent(graph, "RecordEvent", false /* is_forward */, input_args, new_input_names);
+
+  if (send_bw) {
+    // if we have send op in backward pass (at the end of the graph), add the RecordEvent op after that.
```

Comment:
```suggestion
    // if we have Send in backward pass (at the end of the graph), we make sure the RecordEvent happens after that send by adding Send's outputs to RecordEvent's input list.
```
Also, I think you only need one output from `Send`.

Reply:
Send only has one output :) I

---
In reply to: [409739947](https://github.com/microsoft/onnxruntime/pull/3513#discussion_r409739947) [](ancestors = 409739947)